src/misc/cross.cpp

```diff
@@ -16,26 +16,108 @@
  *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
-
-#include "dosbox.h"
 #include "cross.h"
-#include "support.h"
+
 #include <string>
+#include <vector>
+
 #include <limits.h>
 #include <stdlib.h>
+#include <sys/types.h>
+#include <unistd.h>
 
 #ifdef WIN32
 #ifndef _WIN32_IE
 #define _WIN32_IE 0x0400
 #endif
 #include <shlobj.h>
+#else
+#include <libgen.h>
 #endif
 
-#if defined HAVE_SYS_TYPES_H && defined HAVE_PWD_H
-#include <sys/types.h>
+#if defined HAVE_PWD_H
 #include <pwd.h>
 #endif
 
+#include "support.h"
+
+#ifndef WIN32
+
+std::string cached_conf_path;
+
+static bool CreateDirectories(const std::string &path)
+{
+	struct stat sb;
+	if (stat(path.c_str(), &sb) == 0) {
+		const bool is_dir = ((sb.st_mode & S_IFMT) == S_IFDIR);
+		return is_dir;
+	}
+
+	std::vector<char> tmp(path.length());
+	std::copy(path.begin(), path.end(), tmp.begin());
+	std::string dname = dirname(tmp.data());
+
+	// Create parent directories recursively
+	if (!CreateDirectories(dname))
+		return false;
+
+	return (mkdir(path.c_str(), 0700) == 0);
```

Comment:
I'm not sure how to easily solve this, but the `700` doesn't respect the user's UMASK (where they've set it or their distro has set it).

If there's an easy way to determine and use UMASK, then I think we should try - but if it's a mess then then 700 is reasonable. 

Some might argue that a better default would be 755 given the config file holds no personal data -- and I would agree, if it weren't for the fact that we also hold screen captures off this directory; and those can be be highly personal.

Reply:
Edit: scratch this comment.. I think 700 is probably a good default for that last reason, even if the user has set a more generous UMASK. (leaving these comments here only as a basis for these permissions).

Reply:
For the context: 700 was chosen as this is a requirement of XDG spec :)

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

> If, when attempting to write a file, the destination directory is non-existant an attempt should be made to create it with permission 0700. If the destination directory exists already the permissions should not be changed.